Add MSI-X support

sboeuf opened the pull request:
In order to allow virtio-pci devices to use MSI-X messages instead
of legacy pin based intterrupts, this patch implements the MSI-X
support for cloud-hypervisor.

There are three main changes carried by this patch:
- MSI-X support is added to the pci crate. This allows any PCI device
  to be able to leverage MSI-X support.
- virtio-pci has been modified to support MSI-X, based on the code
  provided through the pci crate.
- VirtioDevice trait has been modified to accept a generic closure as
  the interrupt_evt instead of a very specific EventFd. This is an
  important part that let the vmm glue decide what should be done when
  a virtio device wants to send an IRQ to the driver. This would let
  any VMM support multiple types of interrupts, without having to let
  the virtio device know about those.
  In the case of cloud-hypervisor, we don't need to provide the VmFd
  to the virtio device implementation, so that it does not need to
  be aware about KVM specific bits.

Fixes #12 

Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

sboeuf commented:
This PR depends on the pending PR https://github.com/rust-vmm/kvm-ioctls/pull/42

@rbradford @sameo the PR still need a bit of cleanup but I already did most of it, that's why I'd like you to take a look to let me know what you think about the PR.

sameo commented on pci/src/device.rs:
```diff
@@ -41,7 +43,16 @@ impl Display for Error {
-    fn assign_irq(&mut self, _irq_evt: EventFd, _irq_num: u32, _irq_pin: PciInterruptPin) {}
+    fn assign_irq(
+        &mut self,
+        _irq_evt: Arc<Box<Fn(u16) + Send + Sync>>,
```
Would the `Fn(u16)` closure parameter be an IRQ number for non MSI/PCI based transports?

sboeuf replied:
Well in case of IRQ, we will simply ignore it. The point of this `u16` is to select based on the right Queue, but in case of IRQs, there's only one IRQ for all the queues from a device.

sboeuf replied:
Nevermind, I changed it so that it takes no parameter at all.

sameo commented on pci/src/msix.rs:
```diff
@@ -0,0 +1,210 @@
+use byteorder::{ByteOrder, LittleEndian};
+use vm_memory::ByteValued;
+
+const MAX_MSIX_VECTORS_PER_DEVICE: u16 = 0x800;
```
2048

sboeuf replied:
Ok!

sameo commented on pci/src/msix.rs:
```diff
@@ -0,0 +1,210 @@
+        let mut table_entries: Vec<MsixTableEntry> = Vec::new();
+        table_entries.resize_with(msix_vectors as usize, Default::default);
+        let mut pba_entries: Vec<u64> = Vec::new();
+        pba_entries.resize_with(msix_vectors as usize, Default::default);
```
Since the PBA and the MSI-X entries tables will always be the same length, why not having:

```Rust
pub struct Msix {
    pub msg_addr_lo: u32,
    pub msg_addr_hi: u32,
    pub msg_data: u32,
    pub vector_ctl: u32,
}

pub struct MsixTableEntry {
    pub vector: Msix,
    pub pba: u64,
}

pub struct MsixConfig (pub Vec<MsixTableEntry>);

```

sboeuf replied:
They won't. My implementation of PBA is wrong. It's a simple bit array of length "number of vectors". This means in case we have 2 vectors, the PBA will have only 1 entry to cover the 2 bits.
If we had 75 vectors, then we would have 2 entries in the PBA to cover the 75 bits.
I've fixed it now.

sameo commented on vm-virtio/src/transport/pci_device.rs:
```diff
@@ -307,17 +324,43 @@ impl VirtioPciDevice {
+        irq_pin: PciInterruptPin,
+    ) {
         self.configuration.set_irq(irq_num as u8, irq_pin);
         self.interrupt_evt = Some(irq_evt);
```
How am I going to write on my `EventFd` when I'm not using MSI?

sboeuf replied:
Just pass a callback that does it.

sameo commented on vm-virtio/src/net.rs:
```diff
@@ -538,7 +536,7 @@ impl VirtioDevice for Net {
         &mut self,
         mem: GuestMemoryMmap,
-        interrupt_evt: EventFd,
+        interrupt_evt: Arc<Box<Fn(u16) + Send + Sync>>,
```
We may want to rename all those `interrupt_evt` as `interrupt_cb` now.

sboeuf replied:
Sounds good!

sameo commented on vm-virtio/src/device.rs:
```diff
@@ -13,6 +13,8 @@ use std::sync::Arc;
 use vm_memory::GuestMemoryMmap;
 use vmm_sys_util::EventFd;
 
+pub type IrqClosure = Arc<Box<Fn(u16) + Send + Sync>>;
```
I think the closure should be more generic to support non PCI transports. Ideally we would pass a `Queue` clone or reference to it.

sboeuf replied:
Done

sameo commented on vm-virtio/src/block.rs:
```diff
@@ -374,13 +375,11 @@ impl<T: DiskFile> BlockEpollHandler<T> {
-            error!("Failed to signal used queue: {:?}", e);
-            DeviceError::FailedSignalingUsedQueue(e)
-        })
+        (self.interrupt_evt)(self.queues[queue_index].msix_vector);
```
This makes the code PCI/MSI specific. See my comment on `IrqClosure`.

sboeuf replied:
Yes I agree, I'll rework this.

sboeuf replied:
Done

sameo commented on vmm/src/vm.rs:
```diff
@@ -434,16 +439,22 @@ impl DeviceManager {
-            .map_err(DeviceManagerError::Irq)?;
-        // Let's use irq line INTA for now.
-        virtio_pci_device.assign_irq(irqfd, irq_num as u32, PciInterruptPin::IntA);
+        let vm_fd_clone = vm_fd.clone();
```
It seems weird that we have to clone the VM fd and not take a reference on it. The lifetime of the IRQ closure will never exceeds that of the VM and its fd, it seems there should be a way of not cloning it and avoid the kvm-ioctl change.

sboeuf replied:
I used an `Arc<>` instead, works fine!

sboeuf commented on pci/src/device.rs:
```diff
@@ -41,7 +43,16 @@ impl Display for Error {
-    fn assign_irq(&mut self, _irq_evt: EventFd, _irq_num: u32, _irq_pin: PciInterruptPin) {}
+    fn assign_irq(
+        &mut self,
+        _irq_evt: Arc<Box<Fn(u16) + Send + Sync>>,
```
Will do!

sboeuf commented on pci/src/msix.rs:
```diff
@@ -0,0 +1,210 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
```
Yeah I did, I'll replace thanks!

sboeuf commented on pci/src/device.rs:
```diff
@@ -41,7 +43,16 @@ impl Display for Error {
+        _irq_evt: Arc<Box<Fn(u16) + Send + Sync>>,
+        _irq_num: u32,
+        _irq_pin: PciInterruptPin,
+    ) {
```
Agreed 100%. Just wanted to get your feedback on the global idea of using callback here before to clean things up into proper commits.

sboeuf commented:
@sameo @rbradford comments addressed, let me know what you think. In the meantime, I'll split the PR into proper commits.

sboeuf commented:
@sameo @rbradford ok the code is ready now. I think I addressed all comments. Please let me know if I missed anything.
Please be aware that this PR needs to be slightly updated once we get https://github.com/rust-vmm/kvm-ioctls/pull/42 merged.

sameo commented on pci/src/device.rs:
```diff
@@ -3,14 +3,20 @@
+pub type IrqClosure = Box<Fn() -> std::result::Result<(), std::io::Error> + Send + Sync>;
+pub type MsixClosure =
+    Box<Fn(MsixTableEntry) -> std::result::Result<(), std::io::Error> + Send + Sync>;
+
```
We're up to 3 different closure types. For clarity sake, I would suggest defining:

```Rust
pub type PciInterruptClosure =
    Box<Fn(Option<MsixTableEntry>) -> std::result::Result<(), std::io::Error> + Send + Sync>;
```

and use the same type for both `assign_irq` and assign_msix`.

sameo replied:
Also, we need to document the closure prototype.

sboeuf replied:
Regarding your proposal about `PciInterruptClosure`, I'm not sure this is the right approach because those closures should be passed through dedicated functions (`assign_irq` or `assign_msix`) depending on the use case, hence the need to have different closures. In the context of IRQ, why would we have an `Option<MsixTableEntry>` as parameter?

The VMM (`vm.rs`) can setup different closures for its PCI devices, based on the type of interrupt they want to set, this does not need to be generic. On the other hand, `VirtioInterruptClosure` is specific to virtio and has to be generic.

Actually the thing is that `IrqClosure` could be defined somewhere else than the `pci` crate but I'm not sure where.

chao-p replied:
Right, the IrqClosure does not have to be PCI specific and can live in another place. Probably in vm-device once it's ready. But I'm really not a fan of including type as part of type_name, e.g. from type I know this is a closure already.

sameo replied:
> Regarding your proposal about `PciInterruptClosure`, I'm not sure this is the right approach because those closures should be passed through dedicated functions (`assign_irq` or `assign_msix`) depending on the use case,

As a matter of fact, I even considered commenting on the `assign_msix` addition :)
What we're trying to do is extend the idea of assigning an interrupt to a device. Those interrupts can be delivered from the device in several ways: Pin based interrupts or memory based ones.

So I would go further and say we should define a generic `InterruptParameters` structure containing an `Option<MsixTableEntry>` and an `InterruptType` fields.
Then have your closure essentially being defined as:
```Rust
pub type InterruptDelivery =
    Box<Fn(InterruptParameters) -> std::result::Result<(), std::io::Error> + Send + Sync>;
```

And you and Chao are right: This shouldn't be PCI prefixed as it could potentially belong to the device model crate. For now we can keep it here.

>  hence the need to have different closures. In the context of IRQ, why would we have an `Option<MsixTableEntry>` as parameter?
> 
> The VMM (`vm.rs`) can setup different closures for its PCI devices, based on the type of interrupt they want to set, this does not need to be generic.

Yes but otoh the VMM is only trying to assign an interrupt to a device. And it wants to tell the device manager or for now the virtio layer how the device should deliver the interrupt. 

> On the other hand, `VirtioInterruptClosure` is specific to virtio and has to be generic.
> 
> Actually the thing is that `IrqClosure` could be defined somewhere else than the `pci` crate but I'm not sure where.

sameo replied:
Closely related discussion here: https://github.com/rust-vmm/vm-virtio/pull/4

An `DeviceInterrupt` trait seems to be the trend of the discussion, and is something we may want to consider as well.

sameo commented on vm-virtio/src/device.rs:
```diff
@@ -13,6 +13,9 @@ use std::sync::Arc;
 use vm_memory::GuestMemoryMmap;
 use vmm_sys_util::EventFd;
 
+pub type InterruptClosure =
```
Does this have to be a public type?
If it does, should it be `VirtioInterruptClosure` instead?

sboeuf replied:
I think it has to be public since the other `vm-virtio` rust modules (`block`, `net`, `rng` and `transport::pci_device`) need it. WDYT?

About the naming, I like `VirtioInterruptClosure`, I'll make the change.

chao-p commented on pci/src/device.rs:
```diff
@@ -46,6 +49,9 @@ pub trait PciDevice: BusDevice {
     /// The device may write to `irq_evt` to trigger an interrupt.
     fn assign_irq(&mut self, _irq_evt: Arc<IrqClosure>, _irq_num: u32, _irq_pin: PciInterruptPin) {}
```
Once you added msix then I'd change this to assign_pin_irq for clarification. Also _irq_evt to _irq_cb to keep consistent.

sboeuf replied:
Sounds good!

chao-p commented on vm-virtio/src/queue.rs:
```diff
@@ -218,6 +218,9 @@ pub struct Queue {
     pub ready: bool,
 
+    /// MSI-X vector of the queue
+    pub msix_vector: u16,
```
So you are adding some PCI concepts to virtio core. This breaks your assumption that virtio code other than virtio-pci should not aware the existence of PCI. Suggest to change this to 'queue_vector' so even in future we may add queue vector support for mmio transport.

sboeuf replied:
Yeah I know, but the queue needs to contain this information, even if it's ignored in the context of virtio-mmio. I'm okay renaming this. I'd say `vector` is enough since that's already part of the Queue structure.

chao-p commented on vm-virtio/src/transport/pci_device.rs:
```diff
@@ -316,6 +341,16 @@ impl PciDevice for VirtioPciDevice {
+        let msix_config = self.msix_config.clone();
+
+        let cb = Arc::new(Box::new(move |queue: &Queue| {
+            (msi_cb)(msix_config.lock().unwrap().table_entries[queue.msix_vector as usize].clone())
```
Is it possible for table_entries overflow here?

sboeuf replied:
It shouldn't, since the `queue.msix_vector` index is set by the guest, based on the size of the table that we provided through PCI config space.
And if for any reason this happens, at runtime Rust would panic and terminate the VM.

chao-p commented on vmm/src/vm.rs:
```diff
@@ -417,12 +418,13 @@ impl DeviceManager {
-        let mut virtio_pci_device = VirtioPciDevice::new(memory, virtio_device)
-            .map_err(DeviceManagerError::VirtioDevice)?;
+        let mut virtio_pci_device =
+            VirtioPciDevice::new(memory, virtio_device, DEFAULT_MSIX_VEC_NUM)
```
DEFAULT_MSIX_VEC_NUM  is 2, how can we guarantee this is enough for all the supported virtio device types. I mean in future when we add a device that requires more than 2 vectors it's easy for us to forget to change here.

sboeuf replied:
Well for now I'm trying to match the behavior we had with pin based irq, and the virtio spec specifies that 2 vectors is the minimum you must specify for a virtio device.
But in the future, we should consider tweaking this number based on the number of queues we use from our virtio devices if we want better performances.

chao-p commented on vmm/src/vm.rs:
```diff
@@ -435,19 +437,22 @@ impl DeviceManager {
+                pad: [0u8; 12],
+            };
+
+            vm_fd_clone.signal_msi(msi_queue)
```
This ioctl has CAP dependency(KVM_CAP_SIGNAL_MSI) I guess we need to check that somewhere.

sboeuf replied:
Yes I'll add this check, thanks for reminding me :)

rbradford commented on arch/Cargo.toml:
```diff
@@ -6,7 +6,7 @@ authors = ["The Chromium OS Authors"]
 byteorder = "=1.2.1"
 kvm-bindings = "0.1"
-kvm-ioctls = "0.1.0"
+kvm-ioctls = { git = "https://github.com/sboeuf/kvm-ioctls", branch = "kvm_signal_msi" }
```
Obviously fix this before merging

sboeuf replied:
Of course :)

sboeuf commented:
@sameo @rbradford @chao-p I've updated the PR (please review), but the part about vendoring is not right... I spent some time trying to fix the vendoring but nothing works...
See my comment [here](https://github.com/intel/cloud-hypervisor/pull/45#issuecomment-499281038)

rbradford commented:
I've rebased your branch for you to drop the vendoring related change as that is no longer necessary (and also pull in integration testing)

Shall we add a test to check MSI-X is in use? What would that involve?

sboeuf commented:
@rbradford thanks for rebasing and adding the appropriate test :)
LGTM! Ready to be merged.
